cogs/roles.py

```diff
@@ -257,8 +264,238 @@ async def create_role(self, ctx, *, role: Optional[str] = None):
         await ctx.guild.create_role(name=role, reason="Created with Canary")
         await ctx.send("Role created successfully.")
 
-    @commands.command()
-    async def inchannel(self, ctx):
+    def _save_existing_roles(self,
+                             user: discord.Member,
+                             penalty: bool = False):
+        table = "PenaltyUsers" if penalty else "PreviousRoles"
+
+        roles_id = [
+            role.id for role in user.roles
+            if role.name not in ("@everyone", self.penalty_role)
+        ]
+
+        if not roles_id:
+            return
+
+        conn = sqlite3.connect(self.bot.config.db_path)
+        try:
+            c = conn.cursor()
+            # store roles as a string of IDs separated by spaces
+            t = (user.id, " ".join(str(e) for e in roles_id))
+            c.execute(f"REPLACE INTO {table} VALUES (?, ?)", t)
+            conn.commit()
+        finally:
+            conn.close()
+
+    def _fetch_saved_roles(self,
+                           guild,
+                           user: discord.Member,
+                           penalty: bool = False) -> Optional[list]:
+        table = "PenaltyUsers" if penalty else "PreviousRoles"
+
+        conn = sqlite3.connect(self.bot.config.db_path)
+        try:
+            c = conn.cursor()
+            fetched_roles = c.execute(
+                f"SELECT Roles FROM {table} WHERE ID = ?",
+                (user.id, )).fetchone()
+            # the above returns a tuple with a string of IDs separated by spaces
+
+            # Return list of all valid roles restored from the DB
+            #  - filter(None, ...) strips false-y elements
+            return list(
+                filter(None, (guild.get_role(int(role_id))
+                              for role_id in fetched_roles[0].split(" ")
+                              ))) if fetched_roles else None
+
+        finally:
+            conn.close()
+
+    def _has_penalty_role(self, user: discord.Member):
+        penalty_role = utils.get(user.guild.roles, name=self.penalty_role)
+        return penalty_role and next(
+            (r for r in user.roles if r == penalty_role), None) is None
+
+    def _is_in_penalty_table(self, user: discord.Member):
+        conn = sqlite3.connect(self.bot.config.db_path)
+        try:
+            c = conn.cursor()
+            penalty = c.execute("SELECT * FROM PenaltyUsers WHERE ID = ?",
+                                (user.id, )).fetchone()
+            return penalty is None
```

Comment:
doesnt this do the opposite of what it's supposed to. If penality is none, then is_in_penalty_table should return False, not True. Right now, every new user gets the penalty role !

Reply:
🐒

Reply:
678a97a